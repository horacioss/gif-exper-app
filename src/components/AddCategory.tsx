import React, { Dispatch, SetStateAction, useState } from "react";

export interface AddCategoryProps {
  setCategories: Dispatch<SetStateAction<string[]>>;
}

const AddCategory = ({ setCategories }: AddCategoryProps) => {
  const [inputValue, setInputValue] = useState("");

  const handleInputChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    setInputValue(e.target.value);
  };

  const hanleSubmit = (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();

    if (inputValue.trim().length > 2) {
      setCategories((cats) => [inputValue, ...cats]);
      setInputValue("");
    }
  };

  return (
    <form onSubmit={hanleSubmit}>
      <input type="text" value={inputValue} onChange={handleInputChange} />
    </form>
  );
};

export default AddCategory;
