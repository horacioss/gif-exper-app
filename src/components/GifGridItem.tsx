import { image } from "../types/image";

const GifGridItem: React.FC<image> = ({ url, title }) => (
  <div className="card animate__animated animated__fadein">
    <img src={url} alt={title} />
    <p className="title-image">{title}</p>
  </div>
);

export default GifGridItem;
