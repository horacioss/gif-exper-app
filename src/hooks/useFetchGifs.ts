import { useEffect, useState } from "react";
import { getGifs } from "../helpers/getGifs";
import { image } from "../types/image";

interface myState {
  data: image[];
  loading: boolean;
}

export const useFetchGifs = (category: string) => {
  const [state, setState] = useState<myState>({
    data: [],
    loading: true,
  });

  useEffect(() => {
    getGifs(category).then((imgs) => {
      setState({
        data: imgs,
        loading: false,
      });
    });
  });

  return state;
};
