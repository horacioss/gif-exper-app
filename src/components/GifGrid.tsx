import PropTypes from "prop-types";

import { useFetchGifs } from "../hooks/useFetchGifs";
import GifGridItem from "./GifGridItem";

export interface Props {
  category: string;
}

export const GifGrid = ({ category }: Props) => {
  const { data: images, loading } = useFetchGifs(category);

  return (
    <>
      <h3>{category}</h3>
      <div
        style={{ display: "flex", flexDirection: "row", flexWrap: "wrap" }}
        className="animate__animated animated__fadein"
      >
        {loading ? (
          <p className="animate__flash">Loading...</p>
        ) : (
          images.map((img) => {
            return <GifGridItem key={img.id} {...img} />;
          })
        )}
      </div>
    </>
  );
};

GifGrid.propTypes = {
  category: PropTypes.string.isRequired,
};
