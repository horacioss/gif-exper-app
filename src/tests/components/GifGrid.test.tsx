import { shallow } from "enzyme";
import { GifGrid } from "../../components/GifGrid";
import { useFetchGifs } from "../../hooks/useFetchGifs";

jest.mock("../../hooks/useFetchGifs");

describe("test over <GifGrid /> component", () => {
  const category = "Dragon Ball";

  const mockedUseFetchGifs = useFetchGifs as jest.MockedFunction<
    typeof useFetchGifs
  >;

  //let wrapper = shallow(<GifGrid category={category} />);

  //beforeEach(() => {
  //jest.clearAllMocks();
  //wrapper = shallow(<GifGrid category={category} />);
  //});

  test("should render corectly component", () => {
    mockedUseFetchGifs.mockReturnValue({
      data: [],
      loading: true,
    });

    const wrapper = shallow(<GifGrid category={category} />);
    expect(wrapper).toMatchSnapshot();
  });

  test("should show item after data it was loaded by useFetchGifs", () => {
    const gifs = [
      {
        id: "3r48gh8f7g7345t5rfq2",
        title: "Any title",
        url: "https://localhost/any/thing.jpg",
      },
    ];

    mockedUseFetchGifs.mockReturnValue({
      data: gifs,
      loading: false,
    });

    const wrapper = shallow(<GifGrid category={category} />);

    expect(wrapper.find("p").exists()).toBe(false);
    expect(wrapper.find("GifGridItem").length).toBe(gifs.length);
  });
});
