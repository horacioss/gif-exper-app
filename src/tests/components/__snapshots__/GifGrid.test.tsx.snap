// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test over <GifGrid /> component should render corectly component 1`] = `
<Fragment>
  <h3>
    Dragon Ball
  </h3>
  <div
    style={
      Object {
        "display": "flex",
        "flexDirection": "row",
        "flexWrap": "wrap",
      }
    }
  >
    <p
      className="animate__flash"
    >
      Loading...
    </p>
  </div>
</Fragment>
`;
